### ISSUE TITLE
Bundler 2.2.5 does not include `ruby` as a platform by default in the Gemfile.lock

### ISSUE THREAD

**Issue by schneems** [rubygems/rubygems#4280]
Moving the issue mentioned in this comment to it's own issue with a docker reproduction https://github.com/rubygems/rubygems/issues/4269#issuecomment-759597060

## Repro

```
docker run -it --rm ruby:2.7 bash
git clone https://github.com/schneems/bundler_223_fails_deploy
cd bundler_223_fails_deploy
gem install bundler -v 2.2.5
bundle lock --add-platform x86_64-linux
```

This generates a diff:

```
$ git diff
diff --git a/Gemfile.lock b/Gemfile.lock
index ef67b42..56fb6ba 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -5,9 +5,10 @@ GEM

 PLATFORMS
   x86_64-darwin-19
+  x86_64-linux

 DEPENDENCIES
   rake

 BUNDLED WITH
-   2.2.3
+   2.2.5
```

Now if you try to install that Gemfile.lock on a Linux machine:

```
gem uninstall bundler -v 2.2.5
export BUNDLE_WITHOUT='development:test'
export BUNDLE_DEPLOYMENT=1
export BUNDLE_PATH=vendor/bundle
bundle install -j4
```

It will fail when you try to use rake:

```
bundle exec rake -P
bundler: failed to load command: rake (/bundler_223_fails_deploy/vendor/bundle/ruby/2.7.0/bin/rake)
Gem::Exception: can't find executable rake for gem rake. rake is not currently included in the bundle, perhaps you meant to add it to your Gemfile?
  /usr/local/lib/ruby/2.7.0/bundler/rubygems_integration.rb:374:in `block in replace_bin_path'
  /usr/local/lib/ruby/2.7.0/bundler/rubygems_integration.rb:402:in `block in replace_bin_path'
  /bundler_223_fails_deploy/vendor/bundle/ruby/2.7.0/bin/rake:23:in `<top (required)>'
```

This can be fixed by running 

```
unset BUNDLE_DEPLOYMENT
bundle lock --add-platform ruby
Warning: the running version of Bundler (2.1.4) is older than the version that created the lockfile (2.2.5). We suggest you to upgrade to the version that created the lockfile by running `gem install bundler:2.2.5`.
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Writing lockfile to /bundler_223_fails_deploy/Gemfile.lock
export BUNDLE_DEPLOYMENT=1
bundle exec rake -P
```

**Comment by raulpopadineti** [comment]
This is still failing for me after locking the `ruby` platform in my lockfile.

[Here's the gist](https://gist.github.com/raulpopadineti/a44932909c59b67aec093c212cb699f3) for my Gemfile.lock.

Any other things I could try to make the deploy work?

**Comment by raulpopadineti** [comment]
Actually, scratch my last comment. It turns out I wasn't able to deploy due to some issues with the `bootsnap` gem that was fixed once I upgraded it to the latest version.

All is good! Thanks @schneems!!  🍻

**Comment by brunoarueira** [comment]
I've got this issue while deploying one app from the company I work and the workaround fix this issue. Probably the version 2.2.7 has this issue, because our first Gemfile.lock was based on it.

**Comment by schneems** [comment]
It does look like this issue is mitigated with new Gemfile.lock being generated with the latest bundler:

```
$ gem install bundler
Fetching bundler-2.2.15.gem
Successfully installed bundler-2.2.15
Parsing documentation for bundler-2.2.15
Installing ri documentation for bundler-2.2.15
Done installing documentation for bundler after 3 seconds
1 gem installed
⛄ 3.0.0 🚀  ~/Documents/projects/work/pack_attack (main)
$ reproapp
$ mkdir -p /tmp/88c7fe54c92479f6b8103f78de65ba2f; cd /tmp/88c7fe54c92479f6b8103f78de65ba2f
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using rake 13.0.3
Using bundler 2.2.15
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Writing lockfile to /private/tmp/88c7fe54c92479f6b8103f78de65ba2f/Gemfile.lock
Initialized empty Git repository in /private/tmp/88c7fe54c92479f6b8103f78de65ba2f/.git/
[main (root-commit) 8912aac] first
 2 files changed, 16 insertions(+)
 create mode 100644 Gemfile
 create mode 100644 Gemfile.lock
Creating app... done, ⬢ warm-hamlet-26437
https://warm-hamlet-26437.herokuapp.com/ | https://git.heroku.com/warm-hamlet-26437.git
⛄ 3.0.0 🚀  /tmp/88c7fe54c92479f6b8103f78de65ba2f (main)
$ cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    rake (13.0.3)

PLATFORMS
  ruby
  x86_64-darwin-19

DEPENDENCIES
  rake

BUNDLED WITH
   2.2.15
```

That being said, the initial reproduction still fails. I think I would say that `ruby` should always be in the platforms? It's good that this is fixed for newly generated Gemfile.lock, but there are a whole lot of Gemfile.lock files generated between Christmas and now.

**Comment by deivid-rodriguez** [comment]
Hi @schneems!

My idea to fix this and all other issues that have come up on heroku due to using an older version than the one the generated user's lockfiles is to automatically install the `BUNDLED WITH` version of bundler first thing when `bundle install` is run (if not already present), and the re'exec the `bundle install` command using that version of `bundler`.

The rationale, as proved already many times (specially in `frozen` mode), is that the only version of bundler 100% guaranteed to be able to bundle a given lockfile is the one that generated it.

We've been planning to add this improvement for a while anyways, and I believe heroku will greatly benefit from it so it would basically allow you to remove any custom handling of `BUNDLED WITH` and let users use the same version of `bundler` they use locally.

What do you think?

**Comment by schneems** [comment]
Overall I think such behavior would break the ruby buildpack in very deep ways and my strong preference would be to opt-out of it. I mentioned that before over at https://github.com/rubygems/rubygems/issues/4452#issuecomment-802802127.

**Comment by deivid-rodriguez** [comment]
I fail to understand how tweaking our behaviour when the `BUNDLED WITH` section is present would affect heroku at all, given that you explicitly remove the `BUNDLED WITH` section before running `bundle install`.

Precisely because of that, the bundler version locking feature wouldn't fix the problem presented this ticket, because heroku opts out of it by removing the `BUNDLED WITH` section.

I made some changes to the heroku buildpack so that it respects user's choice of bundler here: https://github.com/heroku/heroku-buildpack-ruby/compare/main...deivid-rodriguez:respect_user_choice_of_bundler_version. It's just a quick hack for manual testing, I didn't even run heroku-buildpack-ruby's tests, I only tried it on a toy app. Anybody running into this issue is welcome to try them out. I think `heroku buildpacks:set https://github.com/deivid-rodriguez/heroku-buildpack-ruby.git#respect_user_choice_of_bundler_version` should do the trick.
